pkg/cmd/cli/cmd/rsync/rsync.go

```diff
@@ -229,7 +236,76 @@ func (o *RsyncOptions) Validate() error {
 
 // RunRsync copies files from source to destination
 func (o *RsyncOptions) RunRsync() error {
-	return o.Strategy.Copy(o.Source, o.Destination, o.Out, o.ErrOut)
+	err := o.Strategy.Copy(o.Source, o.Destination, o.Out, o.ErrOut)
+	if err != nil || !o.Watch {
+		return err
+	}
+
+	dirty := false
+	lastChange := time.Now()
+	var watchError error
+	changeLock := sync.Mutex{}
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("error setting up filesystem watcher: %v", err)
+	}
+	defer watcher.Close()
+
+	go func() {
+		for {
+			select {
+			case event := <-watcher.Events:
+				changeLock.Lock()
+				glog.V(5).Infof("filesystem watch event: %s", event)
+				lastChange = time.Now()
+				dirty = true
+				changeLock.Unlock()
+
+				if event.Op&fsnotify.Remove == fsnotify.Remove {
+					watcher.Remove(event.Name)
+				} else {
+					fsnotification.AddRecursiveWatch(watcher, event.Name)
+				}
+			case err := <-watcher.Errors:
+				changeLock.Lock()
+				watchError = fmt.Errorf("error watching filesystem for changes: %v", err)
+				changeLock.Unlock()
+			}
+		}
+	}()
+
+	err = fsnotification.AddRecursiveWatch(watcher, o.Source.Path)
+	if err != nil {
+		// If this fails, it may be necessary to increase available watches on linux
+		// by running 'sudo sysctl fs.inotify.max_user_watches=65536'
+		return fmt.Errorf("error watching source path %s: %v", o.Source.Path, err)
+	}
+
+	for {
+		changeLock.Lock()
+		if watchError != nil {
+			return watchError
```

Comment:
@bparees you will probably say this doesn't matter, since returning from `RunRsync` will eventually terminate the whole program, but just so that we're all on the same page, this leaves the `changeLock` mutex locked forever.

Might be a "good thing" after all, preventing the goroutine we start above from glogging events that might happen after `RunRsync` has returned and before the program terminates.